knowledge-base/adr/0029-customer-header-handling.md

```diff
@@ -0,0 +1,123 @@
+- Merge on structured object means: Ignore casing for the keys and keep the original casing from object with the highest priority.
+  The `mergeLeftIgnoreCase()` method should do the trick.
+- For the parameters do encoding, only on the SDK headers, the others are controlled by the user.
+- Merge the remaining parts (not header, not parameter) of `getAxiosConfigWithDefaults()`, `HttpRequestConfig` and `DestinationHttpRequestConfig`.
+- The priority for this flat merge of the rest is: HttpRequestConfig > DestinationHttpRequestConfig > getAxiosConfigWithDefaults()
+- Use the resulting request config to execute the request.
+
+Note: In the `execute()` method the CSRF token header is also filled.
+This header should consider as SdkHeader from an origin.
+
+### How to make headers distinguishable
+
+We have the same problem in both the `HttpRequestConfig` and `DestinationHttpRequestConfig`.
+Both contain a single `header` property typed `Record<string,any>` or `Record<string,string>` which contains two sources of headers.
+
+- **Option A**: Keep the single property but use a `HeaderValueObject` containing the value but also information on the origin.
+- **Option B**: Add a property `headerSDK` and `headerProperties` to hold different headers.
+
+We would like to keep the SDK internals in option A from the user:
+
+```ts
+interface HeaderValueObject {
+    value:string
+    origin:'Custom'|'DestinationProperty'|'Destination'|'SDK'
+}
+
+//Parameters are handled analogously and are skipped in the methods to be shorted.
+interface ParameterValueObject {
+    value:string
+    origin:'Custom'|'DestinationProperty'|'SDK'
+}
+
+interface ConfigPublic{
+    headers:Record<string,string>
+    parameters:Record<string,string>
+}
+
+/**
+ * @internal
+ */
+interface ConfigInternal{
+    headers:Record<string,HeaderValueObject>
+    parameters:Record<string,ParameterValueObject>
+}
+
+/**
+ * We use this method so the origin of headers is clear. Will be exported but not on root level.
+ * @internal
+ */
+export function executeHttpRequestInternal(config:ConfigInternal){
+    ...
+}
+
+/**
+ * This method is for direct customer use -> provided headers are custom and have high prio.
+ */
+export function executeHttpRequest(config:ConfigPublic){
+    const withObject = headerObjectWithOriginCustom(config) ///
+    executeHttpRequestInternal(withObject)
+}
+```
+
+## Breaking Changes
+
+We investigate the risk of breaking changes:
+
+- Changing the type of the `header` property from `string|any` to something more structured i.e. `string|HeaderValueObject` is possible (Option A).
+- Extending the interface with new properties is possible - the custom ones remain where they are (Option B).
+- Removing the optional `customHeader` argument for the methods in [authentication-header.ts](<(../../packages/connectivity/src/scp-cf/authorization-header.ts)>) is not possible.
+- There is a risk of intrinsic behavior changes.
+
+## Decision
```

Comment:
The user is working on a live system with modified SDK as a work around.
We should apply this and fix the bug on 2.0 branch.
I added this info to the follow up ticket.